### ISSUE TITLE
k6 doesn't like redirects

### ISSUE THREAD

**Issue by ragnarlonn** [loadimpact/k6#132]
"Stopped after 0 redirects" - seems a bit stingy to me!

```
./k6 run 'import http from "k6/http"; export default function() { http.get("http://loadimpact.com"); }'

          /\      |‾‾|  /‾‾/  /‾/   
     /\  /  \     |  |_/  /  / /   
    /  \/    \    |   _  |  /  ‾‾\  
   /          \   |  | \  \ | (_) | 
  / __________ \  |__|  \__\ \___/  Welcome to k6 v0.11.0!

  execution: local
     output: -
     script: [cmdline] (js)

   duration: 0s, iterations: 1
        vus: 1, max: 1

    web ui: http://127.0.0.1:6565/

ERRO[0001] Error: Get https://loadimpact.com/: stopped after 0 redirects     200ms / 0s
    at <native code>
    at request (k6/http.js:176:1166)
    at get (k6/http.js:182:1272)
    at /Users/ragnarlonn/src/go/src/github.com/loadimpact/k6/[cmdline]:1:153
 
      done [==========================================================]      200ms / 200ms

    errors................: 1
    http_req_blocked......: avg=120.31ms, max=120.31ms, med=120.31ms, min=120.31ms, p90=120.31ms, p95=120.31ms
    http_req_connecting...: avg=118.36ms, max=118.36ms, med=118.36ms, min=118.36ms, p90=118.36ms, p95=118.36ms
    http_req_duration.....: avg=240.38ms, max=240.38ms, med=240.38ms, min=240.38ms, p90=240.38ms, p95=240.38ms
    http_req_looking_up...: avg=1.87ms, max=1.87ms, med=1.87ms, min=1.87ms, p90=1.87ms, p95=1.87ms
    http_req_receiving....: avg=92.64µs, max=92.64µs, med=92.64µs, min=92.64µs, p90=92.64µs, p95=92.64µs
    http_req_sending......: avg=106.42µs, max=106.42µs, med=106.42µs, min=106.42µs, p90=106.42µs, p95=106.42µs
    http_req_waiting......: avg=119.89ms, max=119.89ms, med=119.89ms, min=119.89ms, p90=119.89ms, p95=119.89ms
    http_reqs.............: 1
    iterations............: 1
    vus...................: 1
    vus_max...............: 1
```

**Comment by ragnarlonn** [comment]
I see there is an option to change this behaviour, but it would probably be good to allow redirects by default (so, setting "maxRedirects" default value to be e.g. 10).

**Comment by ragnarlonn** [comment]
I also see someone has *tried* to make 10 the default value by setting it in the code that handles command line options. Unfortunately, it doesn't work for some reason.

**Comment by robingustafsson** [comment]
@ragnarlonn @technololigy So I believe I've tracked down this issue. It's because of how the utility function `cliInt64` (same goes for the other cli utility functions) treats default values in combination with the JSON-null helper library (gopkg.in/guregu/null.v3).

The body of `cliInt64` looks like this:
```go
return null.NewInt(cc.Int64(name), cc.IsSet(name))
```

The `cc.IsSet` call will always return `false` unless you explicitly specify the command line option, a default value is not considered "set".

Now the reason for this seemingly only affecting `MaxRedirects`, is that we between the different calls to `opts.Apply(...)` in `run.go` to cascade options setting we make a call to `opts.SetAllValid(true)`, why is not clear to me yet.

This method "validates" **some** options, which seems to mean we accept the defaults unless they've been overridden in the options cascading. It does not however validate `MaxRedirects`, so that option effectively remains `null` (since it's "invalid").

@technololigy Is `opts.SetAllValid(true)` meant to be a accept-defaults-if-not-otherwise-set type of action or am I misreading this? Is the fix to add `MaxRedirects` to `SetAllValid` or to change the behavior of `cliInt64` and friends so the defaults are considered from the start of the options cascade (https://github.com/loadimpact/k6/blob/master/run.go#L271)?

**Comment by ragnarlonn** [comment]
Good work. I thought this would be an easy bug to fix but had a hard time following the execution path, looking at the code. And I don't get either why SetAllValid() is needed :)

**Comment by robingustafsson** [comment]
@technololigy ping!

**Comment by technololigy** [comment]
Ah, yes, the idea was for this to do:

1. Start with defaults
1. Apply each alternative source of configuration in order of priority
1. Apply defaults for anything not overridden

Then there were a lot of changes to that code with not nearly enough unit tests, the entire main package's code is a mess that needs to be restructured…

**Comment by robingustafsson** [comment]
@liclac I've been bitten by this in the past few days, is your work on refactoring the CLI code going to fix this or should I have a go at it?

**Comment by ppcano** [comment]
Duplicated of #369. Closed by 5998d48
